src/Controller/LoginController.php

```diff
@@ -0,0 +1,107 @@
+
+    private OpenIdConfigurationProviderManager $providerManager;
+
+    public function __construct(OpenIdConfigurationProviderManager $providerManager)
+    {
+        $this->providerManager = $providerManager;
+    }
+
+    /**
+     * @Route("/login", name="login")
+     */
+    public function index(Request $request, SessionInterface $session): Response
+    {
+        $authenticationProvider = null;
+
+        // Look for authentication provider (aliased to "role") in query string.
+        if (null !== $request->query->has('role')) {
+            $authenticationProvider = $request->query->get('role');
+        }
+
+        // Look for authentication provider (aliased to "role") in target path query string.
+        if (null === $authenticationProvider) {
+            $targetPath = $this->getTargetPath($request->getSession(), 'main');
+            if (null !== $targetPath) {
+                $queryString = parse_url($targetPath, PHP_URL_QUERY);
+                if (null !== $queryString) {
+                    parse_str($queryString, $targetQuery);
+                    $authenticationProvider = $targetQuery['role'] ?? null;
+                }
+            }
+        }
+
+        // Get authentication provider from cookie.
+        if (null === $authenticationProvider) {
+            $authenticationProviderCookieName = 'authentication_provider';
+            if (null === $request->get('reset-authentication-provider')) {
+                $authenticationProvider = $request->cookies->get($authenticationProviderCookieName);
+            }
+        }
+
+        $rememberProvider = false;
+
+        $formBuilder = $this->createFormBuilder();
+        foreach ($this->providerManager->getProviderKeys() as $key) {
+            $name = 'open_id_connect.login_provider.'.$key;
+            $formBuilder->add('provider_'.$key, SubmitType::class, [
+                'label' => new TranslatableMessage($name, [], 'login'),
+            ]);
+        }
+        $form = $formBuilder
+            ->add('remember_provider', CheckboxType::class, [
+                'label' => new TranslatableMessage('Remember my choice', [], 'login'),
+                'required' => false,
+            ])
+            ->getForm()
+        ;
+
+        $form->handleRequest($request);
+        if ($form->isSubmitted() && $form->isValid()) {
+            foreach ($this->providerManager->getProviderKeys() as $key) {
+                if ($form->get('provider_'.$key)->isClicked()) {
+                    $authenticationProvider = $key;
+                    break;
+                }
+            }
+            $rememberProvider = $form->get('remember_provider')->getData();
+        }
+
+        if (null !== $authenticationProvider) {
+            try {
+                $this->providerManager->getProvider($authenticationProvider);
+                $response = $this->redirectToRoute('itkdev_openid_connect_login',
+                    ['providerKey' => $authenticationProvider]);
```

Comment:
This could also maybe be generalised so you could return the response at a former stage before the form stuff.
And then use it again when handling the submitted form but with the cookie attached.

Reply:
Fixed in https://github.com/itk-dev/naevnssekretariatet/pull/75/commits/5db766241a47a02c9c084f29967348c60539e756